src/core/QCheck2.ml

```diff
@@ -0,0 +1,1868 @@
+    (* 0b011101... *)
+    let s = Bytes.create (length + 2) in
+    Bytes.set s 0 '0';
+    Bytes.set s 1 'b';
+    for i = 0 to length - 1 do
+      Bytes.set s (i+2) (if RS.bool st then '0' else '1')
+    done;
+    Bytes.unsafe_to_string s
+
+  let int32 : int32 t = fun st ->
+    let x = random_binary_string 32 st |> Int32.of_string in
+    Tree.make_primitive (int32_towards 0l) x
+
+  (** TODO Not actually unsigned, fix in https://github.com/c-cube/qcheck/issues/105 *)
+  let ui32 : int32 t = int32
+
+  let int64 : int64 t = fun st ->
+    let x = random_binary_string 64 st |> Int64.of_string in
+    Tree.make_primitive (int64_towards 0L) x
+
+  (** TODO Not actually unsigned, fix in https://github.com/c-cube/qcheck/issues/105 *)
+  let ui64 : int64 t = int64
+
+  let list_size (size : int t) (gen : 'a t) : 'a list t =
+    size >>= fun size ->
+    let rec loop n =
+      if n <= 0
+      then pure []
+      else liftA2 (List.cons) gen (loop (n - 1))
+    in
+    loop size
+
+  let list (gen : 'a t) : 'a list t = list_size nat gen
+
+  let list_repeat (n : int) (gen : 'a t) : 'a list t = list_size (pure n) gen
+
+  let array_size (size : int t) (gen : 'a t) : 'a array t =
+    (list_size size gen) >|= Array.of_list
+
+  let array (gen : 'a t) : 'a array t = list gen >|= Array.of_list
+
+  let array_repeat (n : int) (gen : 'a t) : 'a array t = list_repeat n gen >|= Array.of_list
+
+  let rec flatten_l (l : 'a t list) : 'a list t =
+    match l with
+    | [] -> pure []
+    | gen :: gens -> liftA2 List.cons gen (flatten_l gens)
+
+  let flatten_a (a : 'a t array) : 'a array t =
+    Array.to_list a |> flatten_l >|= Array.of_list
+
+  let flatten_opt (o : 'a t option) : 'a option t =
+    match o with
+    | None -> pure None
+    | Some gen -> opt gen
+
+  let flatten_res (res : ('a t, 'e) result) : ('a, 'e) result t =
+    match res with
+    | Ok gen -> gen >|= Result.ok
+    | Error e -> pure (Error e)
+
+  let shuffle_a (a : 'a array) : 'a array t = fun st ->
+    let a = Array.copy a in
+    for i = Array.length a-1 downto 1 do
+      let j = Random.State.int st (i+1) in
+      let tmp = a.(i) in
+      a.(i) <- a.(j);
+      a.(j) <- tmp;
+    done;
+    Tree.pure a
+
+  let shuffle_l (l : 'a list) : 'a list t = fun st ->
+    let a = Array.of_list l in
+    let _ = shuffle_a a st in
+    Tree.pure (Array.to_list a)
+
+  let shuffle_w_l (l : ((int * 'a) list)) : 'a list t = fun st ->
+    let sample (w, v) =
+      let Tree.Tree (p, _) = float_bound_inclusive 1. st in
+      let fl_w = float_of_int w in
+      (p ** (1. /. fl_w), v)
+    in
+    let samples = List.rev_map sample l in
+    samples
+    |> List.sort (fun (w1, _) (w2, _) -> poly_compare w1 w2)
+    |> List.rev_map snd
+    |> Tree.pure
+
+  let pair (g1 : 'a t) (g2 : 'b t) : ('a * 'b) t = liftA2 (fun a b -> (a, b)) g1 g2
+
+  let triple (g1 : 'a t) (g2 : 'b t) (g3 : 'c t) : ('a * 'b * 'c) t = (g1 >|= (fun a b c -> (a, b, c))) <*> g2 <*> g3
```

Comment:
It is mainly "convention". In general `<$>` (`('a -> 'b) -> 'a t -> 'b t`) and `>|=` (or `>>|`), `'a t -> ('a -> 'b) -> `b t`, are respectively in Functor and in Monad. In this example, I think that `>|=` is more readable. (But it is a personal opinion)

Reply:
Yeah definitely, I suspect there are a few places where this can be cleaned, I was still in the "discovery" phase when I wrote those, I'll take care of it :smile: